worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/segment/SegmentMapPartitionDataReader.java

```diff
@@ -0,0 +1,324 @@
+                "The queued head buffer is not the required segment id, "
+                    + "do not sent it. details: streamId {}, subPartitionId: {}, current segment id: {}, required segment id: {}, reader: {}",
+                streamId,
+                subPartitionId,
+                segmentId,
+                subPartitionRequiredSegmentIds.get(subPartitionId),
+                this);
+            breakLoop = true;
+            break;
+          } else {
+            buffersToSend.poll();
+          }
+        }
+        if (breakLoop) {
+          break;
+        }
+
+        // fetch first buffer and send
+        buffer = fetchBufferToSend();
+        if (buffer == null) {
+          break;
+        }
+        if (buffer instanceof RecyclableSegmentIdBuffer) {
+          logger.warn("Wrong type of buffer, the RecyclableSegmentIdBuffer is not expected.");
+          return;
+        }
+        sendDataInternal(buffer);
+      }
+    }
+
+    tryRecycleReader();
+  }
+
+  @Override
+  protected void addBuffer(ByteBuf buffer, BufferRecycler bufferRecycler) {
+    if (buffer == null) {
+      return;
+    }
+
+    buffer.markReaderIndex();
+    int subPartitionId = buffer.readInt();
+    // check the buffer type
+    boolean isEndOfSegment = isEndOfSegment(buffer, subPartitionId);
+    buffer.resetReaderIndex();
+
+    synchronized (lock) {
+      if (!isReleased) {
+        buffersToSend.add(new RecyclableBuffer(buffer, subPartitionId, bufferRecycler));
+        increaseBacklog();
+      } else {
+        bufferRecycler.recycle(buffer);
+        numInUseBuffers.decrementAndGet();
+        throw new RuntimeException("Partition reader has been failed or finished.", errorCause);
+      }
+
+      subPartitionNextBufferIndex.compute(subPartitionId, (k, v) -> v + 1);
+      if (isEndOfSegment) {
+        updateSegmentId(subPartitionId);
+      }
+    }
+  }
+
+  @Override
+  protected RecyclableBuffer fetchBufferToSend() {
+    synchronized (lock) {
+      if (isReleased || buffersToSend.isEmpty()) {
+        return null;
+      }
+
+      RecyclableBuffer buffer = null;
+      int numCredit = credits.get();
+      if (numCredit > 0) {
+        buffer = buffersToSend.poll();
+      }
+      if (numCredit <= 1) {
+        notifyBacklog(getBacklog());
+      }
+      return buffer;
+    }
+  }
+
+  @Override
+  protected void notifyBacklog(int backlog) {
+    if (backlog == 0) {
+      return;
+    }
+    super.notifyBacklog(backlog);
+  }
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder("SegmentMapDataPartitionReader{");
+    sb.append("startPartitionIndex=").append(startPartitionIndex);
+    sb.append(", endPartitionIndex=").append(endPartitionIndex);
+    sb.append(", streamId=").append(streamId);
+    sb.append('}');
+    return sb.toString();
+  }
+
+  private boolean isEndOfSegment(ByteBuf buffer, int subPartitionId) {
+    boolean isEndOfSegment = false;
+    Preconditions.checkState(
```

Comment:
if `subPartitionId >= startPartitionIndex && subPartitionId <= endPartitionIndex` is false, could this return false instead of checking state?

Reply:
This more like a sanity check. That's to say, if this state is unexpected, there are some bugs in our system that we don't know about yet. IMO, we should not silence the exception by returning false.